### PR TITLE
Make `brew upgrade` and `brew outdated` follow alias changes

### PR DESCRIPTION
**Pull request by penman**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/.github/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).

  <del>I've written <em>a few</em> tests, but plan on writing more.</del> <ins>Done!</ins>
- [x] Have you successfully run `brew tests` with your changes locally?

-----

This pull request makes `brew outdated` and `brew update` follow alias changes, as described in #567.

### CODE REVIEW

File: Library/Homebrew/cmd/outdated.rb [hunk 1]
```diff
@@ -43,15 +43,28 @@ def print_outdated(formulae)
...
           f.pkg_version.to_s
         end
-        puts "#{f.full_name} (#{outdated_versions.join(", ")}) < #{current_version}"
+
+        outdated_versions = outdated_kegs.
+          group_by { |keg| Formulary.from_keg(keg) }.
+          sort_by { |formula, kegs| formula.full_name }.
+          map do |formula, kegs|
+            "#{formula.full_name} (#{kegs.map(&:version) * ", "})"
+          end * ", "
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/outdated.rb]: Can you use a `.join` instead here?

File: Library/Homebrew/formula.rb [hunk 2]
```diff
@@ -1059,39 +1120,77 @@ def migration_needed?
...
   end
 
+  def current_installed_alias_target
+    Formulary.factory(installed_alias_path) if installed_alias_path
+  end
+
+  # Has the target of the alias used to install this formula changed?
+  # Returns false if the formula wasn't installed with an alias.
+  def installed_alias_target_changed?
+    ![self, nil].include?(current_installed_alias_target)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/formula.rb]: Two separate comparisons may make this easier to follow.

File: Library/Homebrew/cmd/outdated.rb (hunk 1 above)
Comment (penman) [inline Library/Homebrew/cmd/outdated.rb]: Sure. Just this one, or the line above too?

Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/outdated.rb]: Both ideally, thanks!

**Comment by MikeMcQuaid** [comment]
@penman Pushed a fix to the CI and my pedantry. Let me know when you think this is good to go and what commands you've used for testing to I can try similarly locally. Thanks!

**Comment by penman** [comment]
From `$(brew --prefix)/Library/Taps/homebrew/homebrew-core`:

```sh
git checkout penman/hello_aliases~1
brew install hello@2
git checkout penman/hello_aliases
brew outdated # hello should not be listed

brew uninstall hello@2
git checkout penman/hello_aliases~1
brew install hello
git checkout penman/hello_aliases
brew outdated # hello should be listed
```

**Comment by penman** [comment]
Here's what I've been using to test. The `penman/hello_aliases` branch is a branch with `hello@2` and `hello@3` formulae, and a `hello` alias. The latest commit changes `hello` from `hello@2` to `hello@3`.

```sh
cd $(brew --prefix)/Homebrew/Library/Taps/homebrew/homebrew-core
git remote add penman https://github.com/penman/homebrew-core.git
git fetch penman
brew uninstall hello # just in case it's installed

git checkout penman/hello_aliases~1
brew install hello@2
brew outdated hello # should be no output
brew outdated hello@2 # should be no output
brew outdated # should not include hello

git checkout penman/hello_aliases
brew outdated hello@2 # should be no output
brew outdated hello # hello should not be listed (because it was installed with a specified version)
brew outdated # hello should not be listed (because it was installed with a specified version)

brew uninstall hello@2
git checkout penman/hello_aliases~1
brew install hello
git checkout penman/hello_aliases
brew outdated hello@2 # should be no output
brew outdated hello # hello should be listed
brew outdated # hello should be listed

brew upgrade hello # should install hello@3
brew outdated hello@2 # should be no output
brew outdated hello@3 # should be no output
brew outdated hello # hello should not be listed
brew outdated # hello should not be listed

brew uninstall hello@2 hello@3 # clean up
```

You can also run the same test commands, but replacing `brew outdated` with `brew upgrade`, and interrupting after seeing what formula are to be upgraded, to check `brew upgrade` has the same behaviour.

**Comment by penman** [comment]
This is getting there, but there are a couple of edge cases I still want to work on.

**Comment by penman** [comment]
> ```sh
> cd $(brew --prefix)/Homebrew/Library/Taps/homebrew/homebrew-core
> git remote add penman https://github.com/penman/homebrew-core.git
> git fetch penman
> brew uninstall hello # just in case it's installed
> 
> git checkout penman/hello_aliases~1
> brew install hello@2
> brew outdated hello # should be no output
> brew outdated hello@2 # should be no output
> brew outdated # should not include hello
> 
> git checkout penman/hello_aliases
> brew outdated hello@2 # should be no output
> brew outdated hello # hello should not be listed (because it was installed with a specified version)
> brew outdated # hello should not be listed (because it was installed with a specified version)
> ```

I'm wondering if what I'm expecting from last two commands is correct. I think maybe it should at least be marked as outdated in `brew outdated hello` because that's specifically asking about this package. That same change would probably make the `brew upgrade hello` equivalent make more sense too (currently it says <samp>hello not installed</samp>).

This is feeling a little like I'm re-implementing pinning, though…

**Comment by MikeMcQuaid** [comment]
> brew outdated hello@2 # should be no output

Agreed.

> brew outdated hello # hello should not be listed (because it was installed with a specified version)

I think this should show as outdated, actually.

> brew outdated # hello should not be listed (because it was installed with a specified version)

Agreed, `hello` should only be listed if you did `brew install hello`.

**Comment by MikeMcQuaid** [comment]
@penman Don't want to put any time pressure on you but ideally 🚢ping this today or tomorrow would be amazing.

**Comment by penman** [comment]
> > brew outdated hello # hello should not be listed (because it was installed with a specified version)

> I think this should show as outdated, actually.

@MikeMcQuaid I just had a look at this, and it'd be pretty difficult to implement. Because it wasn't installed with an alias, there's (currently) no way to know that the `hello` alias used to point to it.

**Comment by MikeMcQuaid** [comment]
@penman Ah, I see. I think it's fine as-is, then. We can always adjust this behaviour and iterate on this if it ends up being confusing in future.

**Comment by penman** [comment]
Okay, as far as I'm concerned this is pretty much ready to go.

The one thing I'm still not sure about is the JSON, as discussed in #567. I'm not too familiar with what this is used for, but I'm still not convinced saying a keg is outdated, but that the current version is the same as the installed version, like we do at the moment makes sense:

```json
[
  {
    "name": "hello",
    "installed_versions": [
      "2.10"
    ],
    "current_version": "2.10"
  }
]
```

I'll leave this up to somebody who's more familiar with it than me, though.

**Comment by MikeMcQuaid** [comment]
> I'll leave this up to somebody who's more familiar with it than me, though.

Will punt on this for now and can iterate in future PRs if there's complaints. Thanks so much for the great work here @penman ✨

**Pull request merged**